### PR TITLE
Adds Docker for macOS, renames docker-engine

### PR DESCRIPTION
**Pull request by mhubig**
#### Editing an existing cask
- [x] Commit message includes cask’s name (and new version, if applicable).
- [x] `brew cask audit --download {{cask_file}}` is error-free.
- [x] `brew cask style --fix {{cask_file}}` left no offenses.
#### Adding a new cask
- [x] Checked there aren’t open [pull requests](https://github.com/caskroom/homebrew-cask/pulls) for the same cask.
- [x] Checked there aren’t closed [issues](https://github.com/caskroom/homebrew-cask/issues) where that cask was already refused.
- [x] When naming the cask, followed the [token reference](https://github.com/caskroom/homebrew-cask/blob/master/doc/cask_language_reference/token_reference.md).
- [x] Commit message includes cask’s name.
- [x] `brew cask audit --download {{cask_file}}` is error-free.
- [x] `brew cask style --fix {{cask_file}}` left no offenses.
- [x] `brew cask install {{cask_file}}` worked successfully.
- [x] `brew cask uninstall {{cask_file}}` worked successfully.
### Some additional background to this pull request
- It renames the recipe for the docker-engine binary from `docker.rb` to `docker-engine.rb` to avoid a naming conflict.
- It adds a recipe ('docker.rb') for the new native Docker for macOS app. It's still in public beta,
  but will eventually replace the legacy Docker Toolbox as the officially Docker application for macOS.

### CODE REVIEW

**Comment by mhubig** [comment]
Closes #22281.

**Comment by mhubig** [comment]
If this gets merged the folks on issue #15603 should be aware of the renaming!

File: Casks/docker.rb [hunk 1]
```diff
@@ -1,20 +1,16 @@
...
+  version '1.12.0.9493'
+  sha256 'cb21128f5c566770325ada5c20b6d6aefdc5e97c993ea816c49a83ecc5e45f85'
 
-  url "https://get.docker.com/builds/Darwin/x86_64/docker-#{version}"
-  appcast 'https://github.com/docker/docker/releases.atom',
-          checkpoint: 'ed067a0c9a4021acb5e59e866192bea0f3dc7e454360a8530b7fa80fe4f0d2b0'
-  name 'Docker Engine Client'
-  homepage 'https://docs.docker.com/engine/userguide/'
-  license :apache
+  # docker.com was verified as official when first introduced to the cask
```
Comment (vitorgalvao) [inline Casks/docker.rb]: Redundant in this case. Please delete the comment.

**Comment by vitorgalvao** [comment]
This now has conflicts.

Instead of `docker` → `docker-engine`, just get rid of the old `docker`. We were going to do that anyway.

**Comment by nixpulvis** [comment]
Should I avoid having both this and `brew install docker` on my system?

I personally don't really like the whole notion of a Cask, because it seems like a pretty artificial distinction to be making about a kind of application, and this is a **perfect** example of where I find it confusing. Don't conflate or alias if at all possible.

```sh
# My install story...

$ brew install docker
...
$ docker version
Docker version 1.12.1, build 23cf638
$ docker run ubuntu /bin/echo "hello world"
docker: Cannot connect to the Docker daemon. Is the docker daemon running on this host?.

$ open http://www.google.com
...wasted time...

$ brew cask install docker
```

Now to make matters much worse for me, I now manage docker updates in the App. A lot of this might just be me being a command line kind of person, but that's what `homebrew` aims to solve. I read the [token reference](https://github.com/caskroom/homebrew-cask/blob/master/doc/cask_language_reference/token_reference.md) but I'm still not really satisfied. Sorry for a bit of a rant, but Docker is an aim at solving some development environment issues, so I figure here's as good a place as any to talk about this.

**Comment by adidalal** [comment]
@nixpulvis https://github.com/caskroom/homebrew-cask/issues/24239
We'll probably make it `docker-app` or something; feel free to leave your input on the linked issue.

**Comment by lox** [comment]
As far as I know there is no plan for Docker for Mac to replace Docker Toolbox.

**Comment by jcrben** [comment]
@nixpulvis I imagine this applies, altho I'm just beginning to explore it myself: https://docs.docker.com/docker-for-mac/docker-toolbox/#/docker-toolbox-and-docker-for-mac-coexistence

**Comment by VoR0220** [comment]
this definitely needs some updating. Currently the cask is failing on Travis CI so far as I can tell.

**Comment by Govinda-Fichtner** [comment]
Is anybody still working on this? Docker4Mac is long out of beta. If nobody else is working on this I would give it a spin....

**Comment by adidalal** [comment]
@Govinda-Fichtner [`docker`](https://github.com/caskroom/homebrew-cask/blob/master/Casks/docker.rb) itself looks relatively updated.

Are you referring to something else?

**Comment by Govinda-Fichtner** [comment]
@adidalal Your are totally right. I've seen it only later that despite this thread there is an up to date version of docker4mac.

**Pull request merged**